### ISSUE TITLE
Feature request: torch.bincount

### ISSUE THREAD

**Issue by davidbau** [pytorch/pytorch#4570]
Feature request: torch GPU bincount support.

numpy.bincount is an essential tool for working with order statistics, histograms, etc. Recently bincount was added to TensorFlow - https://github.com/tensorflow/tensorflow/pull/13813.  I prefer pytorch, and it would be nice to be able to use bincount in pytorch as well.

**Comment by soumith** [comment]
yea we should add this in.

**Comment by davidbau** [comment]
After I opened this, a colleague mentioned to me that the tensorflow bincount is nice but not as much of a speedup as it could be for our application (counting many histograms in parallel). A significant speedup might be obtained for GPUs if there were a bincount(dim=X) version that can do many bincounts in parallel.

**Comment by vishwakftw** [comment]
I believe this can be closed now.

**Comment by sunshineatnoon** [comment]
Hi, Thanks for adding this feature. Is there any documentation for this function? Thanks!

**Comment by chintak** [comment]
@sunshineatnoon, how about:
```python
  >>> import torch
  >>> help(torch.bincount)
```

**Comment by sunshineatnoon** [comment]
@chintak Thanks!

**Comment by edgarriba** [comment]
@soumith @vishwakftw is there any plan to support `torch.bincount` for batched input ? I assume that depending on the input, the output might have different shapes. How is pytorch going to handle this needed cases in the future ? Have other related use cases in mind such as how to represent OOP using torch data structures that is still not clear how have to be addressed. Heard about `__tensor_wrap__`, is this the direction to go ?

**Comment by soumith** [comment]
@edgarriba we haven't thought yet about torch.bincount for batched input. We can just return a tuple of Tensors if we have variable length outputs

**Comment by AranKomat** [comment]
@soumith @edgarriba You can just use scatter_add_(dim, index, other) for batched bincount, where 'index' is your input and 'other' is just torch.ones(...).
